packages/worker/src/worker-options.ts

```diff
@@ -420,6 +426,39 @@ export function addDefaultWorkerOptions(options: WorkerOptions): WorkerOptionsWi
   };
 }
 
+function checkHeapSizeLimit() {
+  if (process.platform === 'linux') {
+    const cgroupMemoryConstraint = Number(
+      tryReadFileSync(/* cgroup v2 */ '/sys/fs/cgroup/memory.max') ??
+        tryReadFileSync(/* cgroup v1 */ '/sys/fs/cgroup/memory/memory.limit_in_bytes')
+    );
+
+    if (cgroupMemoryConstraint < os.totalmem() && cgroupMemoryConstraint < v8.getHeapStatistics().heap_size_limit) {
+      const totalMemInMb = toMB(os.totalmem(), 0);
+      const suggestedOldSpaceSizeInMb = toMB(os.totalmem() * 0.75, 0);
+
+      Runtime.instance().logger.warn(
+        `This program is running inside a containerized environment with a memory constraint ` +
+          `(eg. docker --memory ${totalMemInMb}m). Node itself does not consider this memory constraint ` +
+          `in how it manages its heap memory. There is consequently a high probability that ` +
+          `the process will crash due to running out of memory. To increase reliability, we recommend ` +
+          `adding '--max-old-space-size=${suggestedOldSpaceSizeInMb}' to your node arguments. ` +
+          `Refer to https://docs.temporal.io/application-development/worker-performance for more ` +
+          `advice on tuning your workers.`
+      );
+    }
+  }
+}
+checkHeapSizeLimit();
```

Comment:
Move this to the Runtime constructor so it isn't automatically initialized when its logger is used